client/tinychain/ml/dnn.py

```diff
@@ -12,35 +12,135 @@ def layer(weights, bias, activation):
+            
+            let:-
+            ----
+            ij = (3,5)       3 neurons input layer , 5  neurons Next layer
+            ki = (1000,3)    1000 rows of data , 3 neurons this layer
+                             3,5  * 1000,3 -> 1000,5
+            kj   k = 1000 , j = 5
+            so basically einstein sum here is just simple np.dot(self.inputs,weights)
+            '''
+            
             return activation.forward(einsum("ij,ki->kj", [self.weights, inputs])) + self.bias
 
+
         def gradients(self, A_prev, dA, Z):
+            
+            '''
+            gradients function :it calculates delta values in Weights , Bias and in previous layer.
+            A_prev : Previous Input.
+            Z : Current Output before Activation.
+            dA: detla_Error.
+            dZ : delta Error in the Current Layer.
+            dA_Prev: delta Error in Previous Layer.
+            d_weights: delta Weights.
+            d_bias : delta Bias.
+            activation.backward(dA,Z) : performs derivative version of your activation function
+                                          to apply backward propagation.
+            
+            '''
             dZ = activation.backward(dA, Z).copy()
-            dA_prev = einsum("kj,ij->ki", [dZ, self[0]])
+            dA_prev = einsum("kj,ij->ki", [dZ, self.weights])  
             d_weights = einsum("kj,ki->ij", [dZ, A_prev])
             d_bias = dZ.sum(0)
             return dA_prev, d_weights, d_bias
 
         def train_eval(self, inputs):
+            '''
+            train_eval function you can use it on any layer to get its
+            output after dot product and after applying activation function on that output
+            
+            Z : Output of dot product between weight and inputs .
+            A : returning the output after applying activation function on it.
+            '''
             Z = einsum("ij,ki->kj", [self.weights, inputs])
             A = activation.forward(Z) + self.bias
             return A, Z
 
         def update(self, d_weights, d_bias):
+            '''
+            update function is used to ensure that update operations are done
+            New_Weight = Old Weight - Delta Weight
+            New_Bias = Old Bias - Delta Bias
+            '''
             return self.weights.write(self.weights - d_weights), self.bias.write(self.bias - d_bias)
+                
 
     return DNNLayer([weights, bias])
 
+    
+    """
+    
+    To Build Adam Optimizer or w/e you need just follow these steps:-
+    1. Create Your Class with Your Optimizer Name.
+    2. Build `Create` method as a class method that would initialize everything you need to set before optimizing
+    3. Build `Optimize` method to be called on each layer so it can return delta weights and delta bias also if 
+       you need to update some iterative parameters you might need to map all of them in function return.
+    
+    """
+    
+class Adam_Optimizer():
```

Comment:
Please name this `AdamOptimizer` (without the underscore) for consistency with the PEP-8 style guide: https://www.python.org/dev/peps/pep-0008/#class-names

Reply:
You should also add an abstract superclass to define the API for other Optimizers to follow, and abstract away the implementation details of AdamOptimizer from NeuralNet (or whatever other data structure might benefit from gradient descent).

You can do this with the abc module: https://docs.python.org/3/library/abc.html